src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Definition/RazorDefinitionEndpoint.cs

```diff
@@ -0,0 +1,144 @@
+            _documentResolver = documentResolver ?? throw new ArgumentNullException(nameof(documentResolver));
+            _componentSearchEngine = componentSearchEngine ?? throw new ArgumentNullException(nameof(componentSearchEngine));
+        }
+
+        public TextDocumentRegistrationOptions GetRegistrationOptions()
+        {
+            return new TextDocumentRegistrationOptions
+            {
+                DocumentSelector = RazorDefaults.Selector
+            };
+        }
+
+        public async Task<LocationOrLocationLinks> Handle(DefinitionParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                var path = request.TextDocument.Uri.GetAbsoluteOrUNCPath();
+                _documentResolver.TryResolveDocument(path, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(documentSnapshot.FileKind))
+            {
+                return null;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var originTagHelperBinding = await GetOriginTagHelperBindingAsync(documentSnapshot, codeDocument, request.Position).ConfigureAwait(false);
+            if (originTagHelperBinding is null)
+            {
+                return null;
+            }
+
+            var originTagDescriptor = originTagHelperBinding.Descriptors.FirstOrDefault();
+            if (originTagDescriptor is null)
+            {
+                return null;
+            }
+
+            var originComponentDocumentSnapshot = await _componentSearchEngine.TryLocateComponentAsync(originTagDescriptor).ConfigureAwait(false);
+            if (originComponentDocumentSnapshot is null)
+            {
+                return null;
+            }
+
+            var originComponentUri = new UriBuilder
+            {
+                Path = originComponentDocumentSnapshot.FilePath,
+                Scheme = Uri.UriSchemeFile,
+                Host = string.Empty,
+            }.Uri;
+
+            return new LocationOrLocationLinks(new[]
+            {
+                new LocationOrLocationLink(new Location
+                {
+                    Uri = originComponentUri,
+                    Range = new Range(new Position(0, 0), new Position(0, 0)),
+                }),
+            });
+        }
+
+        public void SetCapability(DefinitionCapability capability)
```

Comment:
Any harm in using a capability property with a private getter?

Reply:
What do you mean by that?

Reply:
https://github.com/dotnet/aspnetcore-tooling/pull/2304/files#diff-ee06127030bb4ab752a7e05d9a118f8fR25

public DefinitionCapability Capability { private get; set; }